### PR TITLE
Fixed Bug in sorting/merge_sort.cpp

### PR DESCRIPTION
**Pull request by ayaankhan98**
#### Description of Change

- Fixed Bug ```munmap_chunck: invalid pointer [core dumped]```
- Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- Relevant documentation/comments is changed or added
- I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: sorting/merge_sort.cpp [hunk 1]
```diff
@@ -1,5 +1,36 @@
+/**
+ *  \addtogroup sorting Sorting Algorithms
+ *  @file{
+ *  \file
+ *  \breif [Merege Sort Algorithm
+ *  (MEREGE SORT)](https://en.wikipedia.org/wiki/Merge_sort) implementation
+ *
+ *  \author [Ayaan Khan] (http://github.com/ayaankhan98)
```
Comment (kvedala) [inline sorting/merge_sort.cpp:8]:
```suggestion
 *  \author [Ayaan Khan](http://github.com/ayaankhan98)
```

File: sorting/merge_sort.cpp [hunk 2]
```diff
@@ -1,5 +1,36 @@
+/**
+ *  \addtogroup sorting Sorting Algorithms
+ *  @file{
```
Comment (kvedala) [inline sorting/merge_sort.cpp:3]:
```suggestion
 *  @{
```

File: sorting/merge_sort.cpp [hunk 3]
```diff
@@ -1,5 +1,36 @@
...
+ *  \file
+ *  \breif [Merege Sort Algorithm
```
Comment (kvedala) [inline sorting/merge_sort.cpp:5]:
```suggestion
 *  \brief [Merege Sort Algorithm
```

File: sorting/merge_sort.cpp [hunk 4]
```diff
@@ -1,5 +1,36 @@
...
+ * sorted sub-arrays into one.
+ *
+ * @param arr[] is the array with two halves one is arr[l...m] and
+ * other is arr[m+1...l]
+ * @param l is the left index of first half array
+ * @param m is the end index of right index of first half array
+ *
+ * (The second array starts form m+1 and goes till l)
+ *
+ * @param l is the end index of right index of second half array
```
Comment (kvedala) [inline sorting/merge_sort.cpp:32]:
```suggestion
 * @param r end or right index of second half array
```

File: sorting/merge_sort.cpp [hunk 5]
```diff
@@ -40,42 +71,47 @@ void merge(int arr[], int l, int m, int r) {
     delete[] R;
 }
 
+/**
+ * Merge sort is a divide and conquer algorithm, it divides the
+ * input array into two halves and calls itself for the two halves
+ * and then calls merge() to merge the two halves
+ *
+ * @param arr[] the array which is to be sorted
+ * @param l define the left index of array
```
Comment (kvedala) [inline sorting/merge_sort.cpp:80]:
```suggestion
 * @param l left index of array
```

File: sorting/merge_sort.cpp [hunk 6]
```diff
@@ -40,42 +71,47 @@ void merge(int arr[], int l, int m, int r) {
...
+ * @param r defines the right index of array
+ *
+ */
 void mergeSort(int arr[], int l, int r) {
```
Comment (kvedala) [inline sorting/merge_sort.cpp:84]:
```suggestion
void mergeSort(int *arr, int l, int r) {
```

File: sorting/merge_sort.cpp [hunk 7]
```diff
@@ -40,42 +71,47 @@ void merge(int arr[], int l, int m, int r) {
...
 }
 
-void show(int A[], int size) {
-    int i;
-    for (i = 0; i < size; i++) std::cout << A[i] << "\n";
+/**
+ * A simple utility function used to print the array after
+ * sorting
+ */
+void show(int arr[], int size) {
```
Comment (kvedala) [inline sorting/merge_sort.cpp:97]:
```suggestion
void show(int *arr, int size) {
```

File: sorting/merge_sort.cpp [hunk 8]
```diff
@@ -1,5 +1,36 @@
...
+ * @param arr[] is the array with two halves one is arr[l...m] and
+ * other is arr[m+1...l]
+ * @param l is the left index of first half array
+ * @param m is the end index of right index of first half array
+ *
+ * (The second array starts form m+1 and goes till l)
+ *
+ * @param l is the end index of right index of second half array
+ */
 void merge(int arr[], int l, int m, int r) {
```
Comment (kvedala) [inline sorting/merge_sort.cpp:34]:
```suggestion
void merge(int *arr, int l, int m, int r) {
```

**Comment by ayaankhan98** [comment]
@kvedala check again, other suggestions are welcome :)

File: sorting/merge_sort.cpp [hunk 9]
```diff
@@ -40,42 +70,47 @@ void merge(int arr[], int l, int m, int r) {
...
-
-    std::cout << "Sorted array\n";
-    show(arr, size);
-
+    mergeSort(arr, 0, size - 1);
+    std::cout << "Sorted array : ";
+    show(arr, size - 1);
     delete[] arr;
     return 0;
 }
```
Comment (kvedala) [inline sorting/merge_sort.cpp:116]:
```suggestion
}
/** @} */
```

Comment (ayaankhan98) [reply sorting/merge_sort.cpp:116]: oh sorry i missed that

**Comment by ayaankhan98** [comment]
can we merge this now? @kvedala

**Comment by kvedala** [comment]
> can we merge this now? @kvedala

yes

**Pull request merged**